python/Ganga/GPIDev/Credentials2/ICredentialInfo.py

```diff
@@ -0,0 +1,192 @@
+from __future__ import absolute_import
+
+import copy
+import os
+from abc import ABCMeta, abstractmethod
+from datetime import datetime, timedelta
+from functools import wraps
+
+import Ganga.Utility.logging
+
+from .exceptions import CredentialsError
+
+logger = Ganga.Utility.logging.getLogger()
+
+
+def cache(method):
+    """
+    The cache decorator can be applied to any method in an ``ICredentialInfo` subclass.
+    It stores the return value of the function in the ``self.cache`` dictionary
+    with the key being the name of the function.
+    The cache is invalidated if ``os.path.getmtime(self.location)`` returns
+    greater than ``self.cache['mtime']``, i.e. if the file has changed on disk
+
+    Not having to call the external commands comes at the cost of calling ``stat()``
```

Comment:
All the credentials that we care about so far (grid proxies and AFS tokens) are stored in `/tmp` so those will be safe. If someone implemented a credential to store, for example, some other username/password in a local file on AFS then I think it should still be fine since AFS supports `mtime`.